lib/srv/heartbeatv2_test.go

```diff
@@ -0,0 +1,325 @@
+	// its non-default value (changed=true/ok=false). Set by tests to trigger
+	// limited traversal of unhappy path.
+
+	pollChanged int
+	fallbackErr int
+	announceErr int
+}
+
+func (h *fakeHeartbeatDriver) Poll() (changed bool) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.pollCount++
+	if h.pollChanged > 0 {
+		h.pollChanged--
+		return true
+	}
+	return false
+}
+
+func (h *fakeHeartbeatDriver) FallbackAnnounce(ctx context.Context) (ok bool) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.fallbackCount++
+	if h.fallbackErr > 0 {
+		h.fallbackErr--
+		return false
+	}
+	return true
+}
+
+func (h *fakeHeartbeatDriver) Announce(ctx context.Context, sender inventory.DownstreamSender) (ok bool) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.announceCount++
+	if h.announceErr > 0 {
+		h.announceErr--
+		return false
+	}
+	return true
+}
+
+func (h *fakeHeartbeatDriver) newStream(ctx context.Context, t *testing.T) client.UpstreamInventoryControlStream {
+	upstream, downstream := client.InventoryControlStreamPipe()
+	select {
+	case h.streamC <- downstream:
+	case <-ctx.Done():
+		t.Fatalf("context canceled during fake stream setup")
+	}
+
+	var msg proto.UpstreamInventoryMessage
+	select {
+	case msg = <-upstream.Recv():
+	case <-ctx.Done():
+		t.Fatalf("context canceled during fake stream recv")
+	}
+
+	_, ok := msg.(proto.UpstreamInventoryHello)
+	require.True(t, ok)
+
+	err := upstream.Send(ctx, proto.DownstreamInventoryHello{
+		ServerID: "test-auth",
+		Version:  teleport.Version,
+	})
+	require.NoError(t, err)
+
+	return upstream
+}
+
+func newFakeHeartbeatDriver() *fakeHeartbeatDriver {
+	// streamC is used to pass a fake control stream to the downstream handle's create func.
+	streamC := make(chan client.DownstreamInventoryControlStream)
+
+	hello := proto.UpstreamInventoryHello{
+		ServerID: "test-node",
+		Version:  teleport.Version,
+		Services: []types.SystemRole{types.RoleNode},
+	}
+
+	handle := inventory.NewDownstreamHandle(func(ctx context.Context) (client.DownstreamInventoryControlStream, error) {
+		// we're emulating an inventory.DownstreamCreateFunc here, but those are typically
+		// expected to return an error if no stream can be acquired. we're deliberately going
+		// with a blocking strategy instead here to avoid dealing w/ backoff that could make the
+		// test need to run longer.
+		select {
+		case <-ctx.Done():
+			return nil, trace.Errorf("context canceled while waiting for next control stream")
+		case stream := <-streamC:
+			return stream, nil
+		}
+	}, hello)
+
+	return &fakeHeartbeatDriver{
+		handle:      handle,
+		streamC:     streamC,
+		pollChanged: 1, // first poll is always true
+	}
+}
+
+// TestHeartbeatV2Basics verifies the basic functionality of HeartbeatV2 under various conditions by
+// injecting a fake driver and attempting to force the HeartbeatV2 into all of its happy and unhappy
+// states while monitoring test events to verify expected behaviors.
+func TestHeartbeatV2Basics(t *testing.T) {
```

Comment:
Can this test be run in parallel? It takes 5-6 seconds on my machine which is a bit heavy for a unit test.